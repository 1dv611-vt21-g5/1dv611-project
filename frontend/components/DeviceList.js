import { useState, useEffect } from 'react'
import _ from 'lodash'
import { Heading, Container, Stack, Flex } from '@chakra-ui/react'

import Device from './Device'
import DeviceSkeletons from './DeviceSkeletons'
import Error from './Error'
import useRequest from 'hooks/useRequest'

const Devices = () => {
  const { data: devices, error } = useRequest('/api/devices', { revalidateOnFocus: false })

  if (error) return <Error />

  return (
    <Container maxW="container.lg">
      <Flex my="2rem">
        <Heading as="h1">My devices</Heading>
      </Flex>
      <Stack spacing="1rem">
        {!devices && (
          <DeviceSkeletons />
        )}
        {devices && devices.map((device, index) => (
          <Device key={index} device={device} />
        ))}
      </Stack>
    </Container>
  )
}

// Exports.
export default Devices
