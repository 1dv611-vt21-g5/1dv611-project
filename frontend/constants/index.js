// TODO: Klargör vilka URLs som gäller för dev, be dem ge access till ak222ye@student.lnu.se med några devices!
const portalHost = 'https://yggio-beta.sensative.net/'

export const statStr = {
  backendHost: 'http://localhost:9999',
  portalHost,
  oauthURL: portalHost + 'oauth'
}
