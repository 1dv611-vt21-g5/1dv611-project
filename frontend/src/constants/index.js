const portalHost = 'https://portal.yggio-sandbox.sensative.net';
export const statStr = {
  backendHost: 'http://localhost:9999',
  portalHost,
  oauthURL: portalHost + '/oauth'
};
