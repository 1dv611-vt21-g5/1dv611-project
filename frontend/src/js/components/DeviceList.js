import React from 'react';

import {subscribe, unsubscribe} from '../actions';
import SubscribeButton from './SubscribeButton';

export default class DeviceList extends React.Component {
  constructor (props) {
    super(props);
    this.state = {devices: props.devices};
  }
  sub = async device => {
    await subscribe(device);
    device.subscribed = true;
    this.setState({devices: this.state.devices});
  }
  unsub = async device => {
    await unsubscribe(device);
    device.subscribed = false;
    this.setState({devices: this.state.devices});
  }
  SubscriptionButton = props => {
    return props.item.subscribed ? (
      <SubscribeButton
        color='danger'
        device={props.item}
        method={this.unsub}>
          Unsubscribe
      </SubscribeButton>
    ) : (
      <SubscribeButton
        color='success'
        device={props.item}
        method={this.sub}>
          Subscribe
      </SubscribeButton>
    );
  }
  makeList = () => {
    return this.props.devices.map((item, idx) => {
      return (
        <li className="deviceLi" key={idx}>
          <span>{item.name}</span>
          <this.SubscriptionButton item={item} />
        </li>
      );
    })
  }

  render () {
    return (
      <ul className="deviceUl">
        {this.makeList()}
      </ul>
    );
  }
};
