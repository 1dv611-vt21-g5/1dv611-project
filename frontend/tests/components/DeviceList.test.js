import React from 'react'
import { rest } from 'msw'
import { setupServer } from 'msw/node'
import { render, waitFor, screen, waitForElementToBeRemoved } from '@testing-library/react'

import { statStr } from '../../constants'

import DeviceList from '../../components/DeviceList'
import testDevices from '../fixtures/testDevices.json'

// mostly adapted from https://testing-library.com/docs/react-testing-library/example-intro
const server = setupServer(
  // här mockar vi och interceptar alla anrop till backenden och ersätter med egen testdata
  rest.get(`${statStr.backendHost}/api/devices`, (req, res, ctx) => {
    return res(ctx.json(testDevices))
  }),
  rest.get(`${statStr.backendHost}/api/subscriptions`, (req, res, ctx) => {
    return res(ctx.json([])) // nolla
  })
)

beforeAll(() => server.listen())
beforeEach(() => server.resetHandlers())
afterAll(() => server.close())

it('loads and displays Loading spinner', async () => {
  // här renderar vi komponenten och väntar INTE på att vår fejkade request ska bli färdig, så den står och laddar
  render(<DeviceList />)
  const loadingText = screen.getByText('Loading...')
  expect(loadingText).not.toBeUndefined()
})

it('loads and displays a list of devices', async () => {
  // här renderar vi komponenten och väntar på att den ska bli färdig så listan dyker upp
  const { container } = render(<DeviceList />)
  await waitFor(() => screen.getByText('Z-Wave Strips Guard 700 Meeting Room'))

  expect(container).toMatchSnapshot()
})