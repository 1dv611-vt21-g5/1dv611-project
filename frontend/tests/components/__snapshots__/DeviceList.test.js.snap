// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loads and displays a list of devices 1`] = `
<div>
  <div
    class="chakra-container css-tflc3m"
  >
    <div
      class="css-1m6bwdk"
    >
      <h1
        class="chakra-heading css-0"
      >
        My devices
      </h1>
    </div>
    <div
      class="chakra-stack css-sekf5y"
    >
      <div
        class="css-1hkhsaz"
      >
        <div
          class="css-70qvj9"
        >
          <p
            class="chakra-text css-1oz30jp"
          >
            Z-Wave Strips Guard 700 Meeting Room
          </p>
          <div
            class="css-17xejub"
          />
          <div
            class="css-ohugnc"
          >
            <p>
              Add loadspinner here!
            </p>
          </div>
        </div>
        <div
          class="css-ohugnc"
        >
          <div
            class="css-15g2oxy"
          >
            <div
              class="css-759u60"
            >
              <svg
                baseProfile="tiny"
                class="chakra-icon css-1c13j1z"
                fill="currentColor"
                focusable="false"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                version="1.2"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M19.064 10.109l1.179-2.387c.074-.149.068-.327-.015-.471-.083-.145-.234-.238-.401-.249l-2.656-.172-.172-2.656c-.011-.167-.104-.317-.249-.401-.145-.084-.322-.09-.472-.015l-2.385 1.18-1.477-2.215c-.186-.278-.646-.278-.832 0l-1.477 2.215-2.385-1.18c-.151-.075-.327-.069-.472.015-.145.083-.238.234-.249.401l-.171 2.656-2.657.171c-.167.011-.318.104-.401.249-.084.145-.089.322-.015.472l1.179 2.386-2.214 1.477c-.139.093-.223.249-.223.416s.083.323.223.416l2.215 1.477-1.18 2.386c-.074.15-.068.327.015.472.083.144.234.238.401.248l2.656.171.171 2.657c.011.167.104.317.249.401.144.083.32.088.472.015l2.386-1.179 1.477 2.214c.093.139.249.223.416.223s.323-.083.416-.223l1.477-2.214 2.386 1.179c.15.073.327.068.472-.015s.238-.234.249-.401l.171-2.656 2.656-.172c.167-.011.317-.104.401-.249.083-.145.089-.322.015-.472l-1.179-2.385 2.214-1.478c.139-.093.223-.249.223-.416s-.083-.323-.223-.416l-2.214-1.475z"
                />
              </svg>
              <p
                class="chakra-text css-1lj023j"
              >
                This device has reported data, click to show!
              </p>
            </div>
          </div>
        </div>
      </div>
      <div
        class="css-1hkhsaz"
      >
        <div
          class="css-70qvj9"
        >
          <p
            class="chakra-text css-1oz30jp"
          >
            Z-Wave Strips Drip 700 Coffee
          </p>
          <div
            class="css-17xejub"
          />
          <div
            class="css-ohugnc"
          >
            <p>
              Add loadspinner here!
            </p>
          </div>
        </div>
        <div
          class="css-ohugnc"
        >
          <div
            class="css-15g2oxy"
          >
            <div
              class="css-759u60"
            >
              <svg
                baseProfile="tiny"
                class="chakra-icon css-1c13j1z"
                fill="currentColor"
                focusable="false"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                version="1.2"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M19.064 10.109l1.179-2.387c.074-.149.068-.327-.015-.471-.083-.145-.234-.238-.401-.249l-2.656-.172-.172-2.656c-.011-.167-.104-.317-.249-.401-.145-.084-.322-.09-.472-.015l-2.385 1.18-1.477-2.215c-.186-.278-.646-.278-.832 0l-1.477 2.215-2.385-1.18c-.151-.075-.327-.069-.472.015-.145.083-.238.234-.249.401l-.171 2.656-2.657.171c-.167.011-.318.104-.401.249-.084.145-.089.322-.015.472l1.179 2.386-2.214 1.477c-.139.093-.223.249-.223.416s.083.323.223.416l2.215 1.477-1.18 2.386c-.074.15-.068.327.015.472.083.144.234.238.401.248l2.656.171.171 2.657c.011.167.104.317.249.401.144.083.32.088.472.015l2.386-1.179 1.477 2.214c.093.139.249.223.416.223s.323-.083.416-.223l1.477-2.214 2.386 1.179c.15.073.327.068.472-.015s.238-.234.249-.401l.171-2.656 2.656-.172c.167-.011.317-.104.401-.249.083-.145.089-.322.015-.472l-1.179-2.385 2.214-1.478c.139-.093.223-.249.223-.416s-.083-.323-.223-.416l-2.214-1.475z"
                />
              </svg>
              <p
                class="chakra-text css-1lj023j"
              >
                This device has reported data, click to show!
              </p>
            </div>
          </div>
        </div>
      </div>
      <div
        class="css-1hkhsaz"
      >
        <div
          class="css-70qvj9"
        >
          <p
            class="chakra-text css-1oz30jp"
          >
            Z-Wave Flower Pot 1
          </p>
          <div
            class="css-17xejub"
          />
          <div
            class="css-ohugnc"
          >
            <p>
              Add loadspinner here!
            </p>
          </div>
        </div>
        <div
          class="css-ohugnc"
        >
          <div
            class="css-15g2oxy"
          >
            <div
              class="css-759u60"
            >
              <svg
                baseProfile="tiny"
                class="chakra-icon css-1c13j1z"
                fill="currentColor"
                focusable="false"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                version="1.2"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M19.064 10.109l1.179-2.387c.074-.149.068-.327-.015-.471-.083-.145-.234-.238-.401-.249l-2.656-.172-.172-2.656c-.011-.167-.104-.317-.249-.401-.145-.084-.322-.09-.472-.015l-2.385 1.18-1.477-2.215c-.186-.278-.646-.278-.832 0l-1.477 2.215-2.385-1.18c-.151-.075-.327-.069-.472.015-.145.083-.238.234-.249.401l-.171 2.656-2.657.171c-.167.011-.318.104-.401.249-.084.145-.089.322-.015.472l1.179 2.386-2.214 1.477c-.139.093-.223.249-.223.416s.083.323.223.416l2.215 1.477-1.18 2.386c-.074.15-.068.327.015.472.083.144.234.238.401.248l2.656.171.171 2.657c.011.167.104.317.249.401.144.083.32.088.472.015l2.386-1.179 1.477 2.214c.093.139.249.223.416.223s.323-.083.416-.223l1.477-2.214 2.386 1.179c.15.073.327.068.472-.015s.238-.234.249-.401l.171-2.656 2.656-.172c.167-.011.317-.104.401-.249.083-.145.089-.322.015-.472l-1.179-2.385 2.214-1.478c.139-.093.223-.249.223-.416s-.083-.323-.223-.416l-2.214-1.475z"
                />
              </svg>
              <p
                class="chakra-text css-1lj023j"
              >
                This device has reported data, click to show!
              </p>
            </div>
          </div>
        </div>
      </div>
      <div
        class="css-1hkhsaz"
      >
        <div
          class="css-70qvj9"
        >
          <p
            class="chakra-text css-1oz30jp"
          >
            LoRa_Strips_3_Raw
          </p>
          <div
            class="css-17xejub"
          />
          <div
            class="css-ohugnc"
          >
            <p>
              Add loadspinner here!
            </p>
          </div>
        </div>
        <div
          class="css-ohugnc"
        >
          <div
            class="css-15g2oxy"
          >
            <div
              class="css-759u60"
            >
              <svg
                baseProfile="tiny"
                class="chakra-icon css-1c13j1z"
                fill="currentColor"
                focusable="false"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                version="1.2"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M19.064 10.109l1.179-2.387c.074-.149.068-.327-.015-.471-.083-.145-.234-.238-.401-.249l-2.656-.172-.172-2.656c-.011-.167-.104-.317-.249-.401-.145-.084-.322-.09-.472-.015l-2.385 1.18-1.477-2.215c-.186-.278-.646-.278-.832 0l-1.477 2.215-2.385-1.18c-.151-.075-.327-.069-.472.015-.145.083-.238.234-.249.401l-.171 2.656-2.657.171c-.167.011-.318.104-.401.249-.084.145-.089.322-.015.472l1.179 2.386-2.214 1.477c-.139.093-.223.249-.223.416s.083.323.223.416l2.215 1.477-1.18 2.386c-.074.15-.068.327.015.472.083.144.234.238.401.248l2.656.171.171 2.657c.011.167.104.317.249.401.144.083.32.088.472.015l2.386-1.179 1.477 2.214c.093.139.249.223.416.223s.323-.083.416-.223l1.477-2.214 2.386 1.179c.15.073.327.068.472-.015s.238-.234.249-.401l.171-2.656 2.656-.172c.167-.011.317-.104.401-.249.083-.145.089-.322.015-.472l-1.179-2.385 2.214-1.478c.139-.093.223-.249.223-.416s-.083-.323-.223-.416l-2.214-1.475z"
                />
              </svg>
              <p
                class="chakra-text css-1lj023j"
              >
                This device has reported data, click to show!
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
